### ISSUE TITLE
[native-image] having to delete a class from my project - delay or allow-incomplete-cp options not helping

### ISSUE THREAD

**Issue by aclement** [oracle/graal#1196]
I've got a spring app `native-image` building (yey!) but I am having to patch spring to make it work. I am finding I have to delete a class completely due to it referencing types not on the classpath (this class is actually never used in the app, it just happens to be in a jar).

This is the problem class - I think it is important that it is an enum: https://github.com/spring-projects/spring-boot/blob/36c1c051b8106371d37f8a747cea0e369a8e5c84/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/jdbc/EmbeddedDatabaseConnection.java

When I run native image it fails with:
```
Fatal error: java.lang.NoClassDefFoundError
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:459)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:289)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:427)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:109)
Caused by: java.lang.NoClassDefFoundError: org/springframework/dao/DataAccessException
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.privateGetMethodRecursive(Class.java:3048)
	at java.lang.Class.getMethod0(Class.java:3018)
	at java.lang.Class.getMethod(Class.java:1784)
	at com.oracle.svm.core.hub.DynamicHub.initEnumConstantsAtRuntime(DynamicHub.java:390)
	at com.oracle.svm.hosted.analysis.Inflation.checkType(Inflation.java:194)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
	at com.oracle.svm.hosted.analysis.Inflation.checkObjectGraph(Inflation.java:136)
	at com.oracle.graal.pointsto.BigBang.checkObjectGraph(BigBang.java:599)
	at com.oracle.graal.pointsto.BigBang.finish(BigBang.java:557)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:685)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:524)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:442)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.ClassNotFoundException: org.springframework.dao.DataAccessException
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
```

Notice that `DataAccessException` is the problem (the ClassNotFoundException there). It is missing, it is deliberately not on my application classpath. Now `EmbeddedDatabaseConnection` is in fact not a class that my app uses in practice.  I can assert that because if I delete that class (and its nested classes) from the jar on the classpath and then run native-image, it all works fine.

There are a couple of references to `DataAccessException` from `EmbeddedDatabaseConnection` - the `isEmbedded(DataSource)` method refers to it inside (but not in the signature), and the inner type `IsEmbedded` has a method `doInConnection` that throws `DataAccessException`.  I'm trying the various options that feel like they might help, both `allow-incomplete-classpath` and `delay-class-initialization-to-runtime` (although not quite sure why the latter would help as there are no references to `DataAccessException` from the static initializer).

If I look at the line in the trace `com.oracle.svm.core.hub.DynamicHub.initEnumConstantsAtRuntime(DynamicHub.java:390)
` it is:
```
            Method values = enumClass.getMethod("values");
```

The generated `values()` method in my enum type does not mention `DataAccessException` but of course `getMethod()` is resolving more than just that method as it searches for it. Am I doing something wrong or is it working as designed, or a limitation? Many thanks for any feedback. I'm going to try hacking out the reference to DataAccessException from inside `isEmbedded()` to see if that helps.

**Comment by aclement** [comment]
I removed the reference in `isEmbedded(DataSource)` to `DataAccessException` and that problem is now gone (so it was that reference) but it then uncovers the next problem. The enum values in the type above actually refer to another type that isn't on the classpath. They value declarations look like this:
```
H2(EmbeddedDatabaseType.H2, "org.h2.Driver",
			"jdbc:h2:mem:%s;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE"),
```
and `EmbeddedDatabaseType` is not on the project classpath. Same exception stack trace as in my first comment, the only difference being `EmbeddedDatabaseType` instead of `DataAccessException`.  Is there perhaps an option to say please completely ignore this class during analysis?
Oh, I'm running on a recent rc17 dev build I compiled myself (I think last friday).

**Comment by vjovanov** [comment]
This comes from the new way we handle enumerations. This was introduced 2 months ago an should not be hard to fix.

**Comment by vjovanov** [comment]
The problem seems to be that this class is seen by analysis in the first place. To find what is the cause we would need a way to reproduce the issue. @aclement could you please show us how to reproduce this?

**Comment by aclement** [comment]
Let me try to share this project

**Comment by aclement** [comment]
ok, I have created https://github.com/aclement/graal-1196

If you clone that project there is a little script `compile.sh` that runs `native-image` against the contents of the `unpack` folder. The `unpack` folder is the simple spring app built, unpacked and patched to be graal ready. I am sorry it isn't a minimal testcase but hopefully it reliably fails at least. Thanks for taking a look.

**Comment by vjovanov** [comment]
We are working on a fix, thanks for making it reproducible. Note that on master you will need to add `--initialize-at-build-time`, or even better `--initialize-at-build-time=<csv-of-packages>`.

If you want all your classes delayed to run time you will need:
```
--initialize-at-build-time=org.springframework.util.unit.DataSize,org.springframework.core.annotation.AnnotationFilter,org.springframework.core.annotation.PackagesAnnotationFilter,org.springframework.util.StringUtils,org.springframework.util.Assert
```

In the second case, the list is necessary as lambdas and annotations get initialized at build-time. In Spring those pull in the given classes. We are working on a fix for that.

**Comment by pejovica** [comment]
This is fixed by 78fcee1492c3c8423ff41fced7e802611cedae73.
